### PR TITLE
Unit 4 - CLIP and relatives - Losses - Intro 

### PR DESCRIPTION
**Pull request by pedrogengo**
This section is part of showcasing of CLIP relatives and it covers an introduction to contrastive learning and some important losses like contrastive loss, triplet loss, etc.

The whole section structure:
- Contrastive Learning @pedrogengo
- Contrastive loss @pedrogengo
- Triple Loss @pedrogengo

### CODE REVIEW

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+Before diving into the diffrent losses used to train models like CLIP, it is important to have a clear understanding about what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space, and dissimilar items are distinctly separated.
+
+In image below, we have an example where we want to keep the representation from dogs closer to other dogs, but also far from cats.
+![image info](./contrastive_learning.png)
+
+# Losses
+
+## Contrastive loss
+
+Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.
```
Comment (ricard-inho) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:12]:
```suggestion
Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a pair of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+Explaining in simple terms:
+
+- If the samples are similar ($y_i = y_j$), then we minimize the term $||x_i - x_j||^2$ that corresponds to their Euclidean distance, i.e., we want to make them closer;
+- If the samples are dissimilar ($y_i \neq y_j$), then we minimize the term $max(0, \epsilon - ||x_i - x_j||^2)$ that is equivalent to maximizing their euclidean distance until some limit $\epsilon$, i.e., we want to make them distant from each other.
+
+
+# References
+
+- https://lilianweng.github.io/posts/2021-05-31-contrastive/
+- https://www.baeldung.com/cs/contrastive-learning
```
Comment (ricard-inho) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:27]:
```suggestion
- [An Introduction to Contrastive Learning](https://www.baeldung.com/cs/contrastive-learning)
- [Contrastive Representation Learning](https://lilianweng.github.io/posts/2021-05-31-contrastive/)
```

**Review by ricard-inho — COMMENTED** [review]
So far looks good. I see there's a plan for Triple Loss too, will you be able to add other losses like N-pair Loss or Lifted Structured Loss?

**Comment by pedrogengo** [comment]
> So far looks good. I see there's a plan for Triple Loss too, will you be able to add other losses like N-pair Loss or Lifted Structured Loss?

Yes! We should add this week :)

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Contrastive Learning
+
+Before diving into the diffrent losses used to train models like CLIP, it is important to have a clear understanding about what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space, and dissimilar items are distinctly separated.
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:3]:
```suggestion
Before diving into the different losses used to train models like CLIP, it is important to have a clear understanding of what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space, and dissimilar items are distinctly separated.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+In image below, we have an example where we want to keep the representation from dogs closer to other dogs, but also far from cats.
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:5]:
```suggestion
In the image below, we have an example where we want to keep the representation from dogs closer to other dogs, but also far from cats.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx (hunk 1 above)
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:12]:
```suggestion
Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples close and dissimilar distant.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+$$L = \mathbb{1}[y_i = y_j]||x_i - x_j||^2 + \mathbb{1}[y_i \neq y_j]max(0, \epsilon - ||x_i - x_j||^2)$$
+
+Explaining in simple terms:
+
+- If the samples are similar ($y_i = y_j$), then we minimize the term $||x_i - x_j||^2$ that corresponds to their Euclidean distance, i.e., we want to make them closer;
+- If the samples are dissimilar ($y_i \neq y_j$), then we minimize the term $max(0, \epsilon - ||x_i - x_j||^2)$ that is equivalent to maximizing their euclidean distance until some limit $\epsilon$, i.e., we want to make them distant from each other.
+
+
+# References
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:24]:
```suggestion
## References
```

**Review by mmhamdy — COMMENTED** [review]
Looking good 👍. I just have two comments:

1. I'm just a little confused about the structure. So, this chapter is about losses, and that is its title, right? But the main heading (Heading 1) here is contrastive learning. The main heading here should be losses, in my opinion.

2. I see that you're including images with the files here. I suggest that you upload the image to the storage provided by HF and reference it here To keep the repository itself as lightweight as possible.

**Comment by pedrogengo** [comment]
> Looking good 👍. I just have two comments:
> 
> 1. I'm just a little confused about the structure. So, this chapter is about losses, and that is its title, right? But the main heading (Heading 1) here is contrastive learning. The main heading here should be losses, in my opinion.
> 2. I see that you're including images with the files here. I suggest that you upload the image to the storage provided by HF and reference it here To keep the repository itself as lightweight as possible.

Fair points.

About 1: I agree with you, but I just want to contextualize a bit the umbrella where all these losses are included. I can change this to introduction, or something else.
About 2: Just uploaded the image for review. Im waiting my approval on the org to include the image there

**Comment by mmhamdy** [comment]
> About 1: I agree with you, but I just want to contextualize a bit the umbrella where all these losses are included. I can change this to introduction, or something else.

I see. But isn't this chapter about different losses (not just contrastive ones), I checked the outline and found that you are also planning to write about non-contrastive and triplet losses. My opinion, assuming content about other losses will be added here later, is to use "contrastive loss" as the heading, and within it introduce a short brief about contrastive learning.

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx (hunk 3 above)
Comment (merveenoyan) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:3]:
```suggestion
Before diving into the diffrent losses used to train models like CLIP, it is important to have a clear understanding about what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space and dissimilar items are distinctly separated.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Contrastive Learning
+
+Before diving into the diffrent losses used to train models like CLIP, it is important to have a clear understanding about what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space, and dissimilar items are distinctly separated.
+
+In image below, we have an example where we want to keep the representation from dogs closer to other dogs, but also far from cats.
+![image info](./contrastive_learning.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:6]: You can put your images in hf-vision/course-assets hugging face repository and put the image from here

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx (hunk 1 above)
Comment (merveenoyan) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:12]:
```suggestion
Contrastive loss is one of the first training objectives used for contrastive learning. It takes a pair of samples as input that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+# Losses
+
+## Contrastive loss
+
+Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.
+
+Technically speaking, imagine that we have a list of input samples $\{x_n\}$ from multiple classes. What we want is a function where examples from the same class have their embeddings close in the embedding space, and also examples from different classes are far apart. Translating this to a mathematical equation, what we have is:
```
Comment (merveenoyan) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:14]:
```suggestion
Technically speaking, imagine that we have a list of input samples $\{x_n\}$ from multiple classes. We want a function where examples from the same class have their embeddings close in the embedding space, and examples from different classes are far apart. Translating this to a mathematical equation, what we have is:
```

**Review by merveenoyan — COMMENTED** [review]
Thank you!

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx (hunk 1 above)
Comment (pedrogengo) [reply chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:12]: I just wanted to emphasize that this loss is pair wise haha but I already made it lower case

Comment (MKhalusova) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx:12]: Perhaps, the sentence needs rephrasing. E.g.:

"Contrastive loss is one of the first training objectives used for contrastive learning. It takes a pair of samples as input that can be similar or dissimilar, and the objective is to map similar samples close in the embedding space and push dissimilar samples apart."

**Review by MKhalusova — APPROVED** [review]
LGTM, left a minor suggestion.

**Comment by pedrogengo** [comment]
Addressed the comments :) I'm assigning the other losses to myself, so I will try to write about them this week in order to close this chapter as soon as possible.

**Review by mmhamdy — APPROVED** [review]
Amazing! Looking great 👍

**Comment by merveenoyan** [comment]
Can you either remove the empty files or if those files are updated on upstream, can you update your upstream? Also you can fix Mulitmodal to Multimodal in file name, this is fixed. We can merge then

**Comment by mattmdjaga** [comment]
@merveenoyan is this PR fine for merging?

**Comment by mattmdjaga** [comment]
> hey @mattmdjaga if you can remove: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/Introduction.md and notebooks/Unit 4 - Mulitmodal Models/CLIP and relatives/README.md and also fix the folder name (Mulitmodal -> Multimodal) we can merge this :)

Should be good now.

**Review by ratan — APPROVED** [review]
LGTM

**Pull request merged**